salt/metalk8s/volumes/prepared/init.sls

```diff
@@ -3,51 +3,78 @@ include:
-{%- if target_volume is not none %}
-  {%- if target_volume in all_volumes.keys() %}
+{%- if target_volume_name is not none %}
+  {%- set target_volume = all_volumes.get(target_volume_name) %}
+  {%- if target_volume is not none %}
     {%- do volumes_to_create.append(target_volume) %}
   {%- else %}
-Volume {{ target_volume }} not found in pillar:
+Volume {{ target_volume_name }} not found in pillar:
   test.configurable_test_state:
-    - name: {{ target_volume }}
+    - name: {{ target_volume_name }}
     - changes: False
     - result: False
-    - comment: Volume {{ target_volume }} not found in pillar
+    - comment: Volume {{ target_volume_name }} not found in pillar
   {%- endif %}
 {%- else %}
-    {%- do volumes_to_create.extend(all_volumes.keys()|list) %}
+    {%- do volumes_to_create.extend(all_volumes.values()|list) %}
 {%- endif %}
 
 Create the sparse file directory:
   file.directory:
     - name: /var/lib/metalk8s/storage/sparse/
     - makedirs: True
 
+Set up systemd template unit for sparse loop device provisioning:
+  file.managed:
+    - name: /etc/systemd/system/metalk8s-sparse-volume@.service
+    - source: salt://{{ slspath }}/files/metalk8s-sparse-volume@.service
+    - user: root
+    - group : root
+    - mode: 644
+    - makedirs: True
+    - dir_mode: 755
+
+Install clean-up script:
+  file.managed:
+    - name: /usr/local/bin/metalk8s-volume-clean-up
+    - source: salt://{{ slspath }}/files/metalk8s-volume-clean-up
+    - user: root
+    - group : root
+    - mode: 755
+    - makedirs: True
+    - dir_mode: 755
+
 {%- for volume in volumes_to_create %}
+  {%- set volume_name = volume.metadata.name %}
 
-Create backing storage for {{ volume }}:
+Create backing storage for {{ volume_name }}:
   metalk8s_volumes.present:
-    - name: {{ volume }}
+    - name: {{ volume_name }}
     - require:
       - file: Create the sparse file directory
 
-Prepare backing storage for {{ volume }}:
+Prepare backing storage for {{ volume_name }}:
   metalk8s_volumes.prepared:
-    - name: {{ volume }}
+    - name: {{ volume_name }}
     - require:
       - metalk8s_package_manager: Install e2fsprogs
       - metalk8s_package_manager: Install xfsprogs
       - metalk8s_package_manager: Install gdisk
-      - metalk8s_volumes: Create backing storage for {{ volume }}
+      - metalk8s_volumes: Create backing storage for {{ volume_name }}
+  {%- if 'sparseLoopDevice' in volume.spec %}
 
-Provision backing storage for {{ volume }}:
-  metalk8s_volumes.provisioned:
-    - name: {{ volume }}
+Provision backing storage for {{ volume_name }}:
+  service.running:
+    - name: metalk8s-sparse-volume@{{ volume.metadata.uid }}
+    - enable: true
     - require:
-      - metalk8s_volumes: Prepare backing storage for {{ volume }}
+      - metalk8s_volumes: Prepare backing storage for {{ volume_name }}
+    - watch:
+      - file: Set up systemd template unit for sparse loop device provisioning
```

Comment:
Shouldn't this be a `require` instead of a `watch`? If it needs to be a `watch`, why exactly? We definitely don't want to 'restart' the service if for whatever reason we update the service template, because restarting will likely not be possible (since the loop device is in use by some K8s container).

Reply:
I thought `losetup --find` was idempotent. Using `watch` was indeed to reload the service in case the template changed (e.g. fixing a bug somewhere in that chain), but maybe we shouldn't do this at all.

Reply:
It's not:
```
[root@containerd ~]# losetup -l
[root@containerd ~]# losetup --find /root/disk
[root@containerd ~]# losetup -l
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /root/disk
[root@containerd ~]# losetup --find /root/disk
[root@containerd ~]# losetup -l
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /root/disk
/dev/loop1         0      0         0  0 /root/disk
[root@containerd ~]# losetup --find /root/disk
[root@containerd ~]# losetup -l
NAME       SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE
/dev/loop0         0      0         0  0 /root/disk
/dev/loop1         0      0         0  0 /root/disk
/dev/loop2         0      0         0  0 /root/disk
```

Furthermore, a restart of the service would imply an `losetup -d` of the existing volume (unit is stopped), which we don't want to happen.

Reply:
Used a `require` instead.